python/kvikio/zarr.py

```diff
@@ -15,49 +17,86 @@
+    when applicable.
 
-    TODO: Write metadata to disk in order to preserve the item types such that
-    GPU items are read as GPU device buffers and CPU items are read as bytes.
+    Notes
+    -----
+    GDSStore doesn't implement `_fromfile()` thus non-array data such as
+    meta data is always read into host memory.
+    This is because only zarr.Array use getitems() to retrieve data.
     """
 
+    def __init__(self, *args, **kwargs) -> None:
+        if not kvikio.zarr.supported():
+            raise RuntimeError("GDSStore requires Zarr v2.15+")
+        super().__init__(*args, **kwargs)
+
     def __eq__(self, other):
         return isinstance(other, GDSStore) and self.path == other.path
 
-    def _fromfile(self, fn):
-        """Read `fn` into device memory _unless_ `fn` refers to Zarr metadata"""
-        if os.path.basename(fn) in [
-            zarr.storage.array_meta_key,
-            zarr.storage.group_meta_key,
-            zarr.storage.attrs_key,
-        ]:
-            return super()._fromfile(fn)
-        else:
-            nbytes = os.path.getsize(fn)
-            with kvikio.CuFile(fn, "r") as f:
-                ret = cupy.empty(nbytes, dtype="u1")
-                read = f.read(ret)
-                assert read == nbytes
-                return ret
-
     def _tofile(self, a, fn):
-        a = asarray(a)
-        assert a.contiguous
-        if a.cuda:
-            with kvikio.CuFile(fn, "w") as f:
-                written = f.write(a)
-                assert written == a.nbytes
-        else:
-            super()._tofile(a.obj, fn)
+        with kvikio.CuFile(fn, "w") as f:
+            written = f.write(a)
+            assert written == a.nbytes
+
+    def getitems(
+        self,
+        keys: Sequence[str],
+        *,
+        contexts: Mapping[str, Mapping] = {},
+    ) -> Mapping[str, Any]:
+        default_meta_array = numpy.empty(())
+        files = []
+        ret = {}
+        io_results = []
+        try:
+            for key in keys:
+                filepath = os.path.join(self.path, key)
+                if not os.path.isfile(filepath):
+                    continue
+                try:
+                    meta_array = contexts[key]["meta_array"]
+                except KeyError:
+                    meta_array = default_meta_array
+
+                nbytes = os.path.getsize(filepath)
+                f = kvikio.CuFile(filepath, "r")
+                files.append(f)
+                ret[key] = numpy.empty_like(meta_array, shape=(nbytes,), dtype="u1")
+                io_results.append((f.pread(ret[key]), nbytes))
+
+            for future, nbytes in io_results:
+                nbytes_read = future.get()
+                if nbytes_read != nbytes:
+                    raise RuntimeError(
+                        f"Incomplete read ({nbytes_read}) expected {nbytes}"
+                    )
+        finally:
+            for f in files:
+                f.close()
```

Comment:
The idiomatic way to do this is...

```
import contextlib

...
with contextlib.ExitStack() as stack:
    for key in keys:
        f = stack.enter_context(kvikio.CuFile(filepath, "r"))
        ret[key] = ..
        io_results.append((f.pread(ret[key]), nbytes)
   for fut, nbytes in io_results:
        nbytes_read = fut.get()
        ...
```

Reply:
I'll note that both the new code (and my proposed approach) leave dangling futures if the block produces an exception anywhere. I guess that is fine because that propagates and the program is hosed at that point anyway?